doc/man3/OSSL_PARAM.pod

```diff
@@ -219,6 +237,14 @@ requested data, it must set C<return_size> for each such
 C<OSSL_PARAM> item to the required size, and eventually return an
 error.
 
+=item *
+
+Functions returning a constant C<OSSL_PARAM> array as a parameter
+descriptor may set C<data> and C<return_size> for their own purposes.
+And example would be to have C<data> point at handler functions for
```

Comment:
Ugh.  Let us not suggest this travesty.
Data should either be NULL or point to what a query would receive (e.g. a name or version string).

Reply:
Actually, I was thinking of using this to avoid duplication (which is a maintenance burden), i.e. something like this in a provider:

``` C
typedef (get_fn)(OUTCTX *, const OSSL_PARAM);

int get_foo(OURCTX *ctx, const OSSL_PARAM *p)
{
    return OSSL_PARAM_get_int(p, &ctx->foo);
}

int get_bar(OURCTX *ctx, const OSSL_PARAM *p)
{
    return OSSL_PARAM_get_utf8_ptr(p, &ctx->bar);
}

const OSSL_PARAM known_set_params[] = {
    OSSL_PARAM_int("foo", &get_foo),
    OSSL_PARAM_utf8_ptr("bar", get_bar),
    OSSL_PARAM_END
};

const OSSL_PARAM *our_set_param_types(void)
{
    return known_set_params;
}

int our_set_params(void *vctx, const OSSL_PARAM *params)
{
    const OSSL_PARAM *k;

    for (k = known_params; k->key != NULL; k++) {
        const OSSL_PARAM *p = OSSL_PARAM_locate_const(params, k->key);
        get_fn fn = (get_fn *)k->data;

        if (p != NULL && !fn(vctx, p))
            return 0;
    }
    return 1;
}
```

The alternative is this:

``` C
const OSSL_PARAM known_params[] = {
    OSSL_PARAM_int("foo", NULL),
    /* OSSL_PARAM_utf8_ptr isn't practical with NULL */
    { "bar", OPENSSL_PARAM_UTF8_PTR, NULL, 0, 0 },
    OSSL_PARAM_END
};

const OSSL_PARAM *our_set_param_types(void)
{
    return known_params;
}

int our_set_params(void *vctx, const OSSL_PARAM *params)
{
    OURCTX *ctx = vctx;
    const OSSL_PARAM *p;

    p = OSSL_PARAM_locate_const(params, "foo");
    if (p != NULL && !OSSL_PARAM_int(p, &ctx->foo))
        return 0;
    p = OSSL_PARAM_locate_const(params, "bar");
    if (p != NULL && !OSSL_PARAM_utf8_ptr(p, &ctx->bar))
        return 0;
    return 1;
}
```

I've already had a few occasions when I changed `our_set_params` and forgot to update `known_set_params` accordingly.  That's the maintenance burden...

Reply:
Note that this suggestion leaves it to the provider author the liberty to code in whatever way they find practical.  It doesn't *enforce* this kind of hackery.

Reply:
However, if you prefer, I can simply leave this out, and people will have to figure it out on their own...

Reply:
I see the point and it does make decoding easier.  It's heading towards a table based decoding mechanism i suggested in another PR (to avoid the  repetitive switch statement's case bodies).

My suggestion was for them to hold the current value of the setting which would also be useful to an application.

There are merits for both but I'd not want to see them being mixed.

Reply:
I think it best to not do this at the moment.  I'm ready to appove this PR and this is the stopper for me.

It's a neat idea but let's see where this evolves first.

Reply:
Ok,, I'll simply remove that paragraph.

Reply:
Done